src/libsyntax/parse/parser.rs

```diff
@@ -680,12 +693,62 @@ impl<'a> Parser<'a> {
         }
     }
 
+    fn recover_closing_delimiter(
+        &mut self,
+        tokens: &[token::Token],
+        mut err: DiagnosticBuilder<'a>,
+    ) -> PResult<'a, bool> {
+        let mut pos = None;
+        // we want to use the last closing delim that would apply
+        for (i, unmatched) in self.unclosed_delims.iter().enumerate().rev() {
+            if tokens.contains(&token::CloseDelim(unmatched.expected_delim))
+                && Some(self.span) > unmatched.unclosed_span
+            {
+                pos = Some(i);
+            }
+        }
+        match pos {
+            Some(pos) => {
+                // Recover and assume that the detected unclosed delimiter was meant for
+                // this location. Emit the diagnostic and act as if the delimiter was
+                // present for the parser's sake.
+
+                 // Don't attempt to recover from this unclosed delimiter more than once.
+                let unmatched = self.unclosed_delims.remove(pos);
+                let delim = TokenType::Token(token::CloseDelim(unmatched.expected_delim));
+
+                 // We want to suggest the inclusion of the closing delimiter where it makes
+                // the most sense, which is immediately after the last token:
+                //
+                //  {foo(bar {}}
+                //      -      ^ help: `)` may belong here
```

Comment:
unfortunately the output does not look like this, because (I think) the span of the main diagnostic message is not the same (the main span points to the `}`, while the suggestion points to the space between the `}` and the character before that.

We fixup the emitter and make it treat those situations the same (for single char spans)

Reply:
I don't think this is a high priority issue, but you're right in your assessment of the problem and the cause. It'll probably be a small ticket/PR to fix it.

Reply:
oh, definitely. We should still keep the comment in sync with the docs. (and add a FIXME with the corresponding issue number

Reply:
Filed https://github.com/rust-lang/rust/issues/58270